src/Databases/DatabaseOrdinary.cpp

```diff
@@ -231,54 +130,128 @@ void DatabaseOrdinary::loadTablesMetadata(ContextPtr local_context, ParsedTables
+    LoadingStrictnessLevel /*mode*/)
+{
+    std::scoped_lock lock(mutex);
 
-    auto startup_one_table = [&](const StoragePtr & table)
+    /// Initialize progress indication on the first call
+    if (total_tables_to_startup == 0)
     {
-        /// Since startup() method can use physical paths on disk we don't allow any exclusive actions (rename, drop so on)
-        /// until startup finished.
-        auto table_lock_holder = table->lockForShare(RWLockImpl::NO_QUERY, getContext()->getSettingsRef().lock_acquire_timeout);
-        table->startup();
-        logAboutProgress(log, ++tables_processed, total_tables, watch);
-    };
+        total_tables_to_startup = tables.size();
+        startup_watch.restart();
+    }
 
+    auto job = makeLoadJob(
+        std::move(startup_after),
+        AsyncLoaderPoolId::BackgroundStartup,
+        fmt::format("startup table {}", name.getFullName()),
+        [this, name] (AsyncLoader &, const LoadJobPtr &)
+        {
+            // `DatabaseOnDisk::tryGetTable()` is used to avoid waiting for a startup
+            if (auto table = DatabaseOnDisk::tryGetTable(name.table, {}))
+            {
+                /// Since startup() method can use physical paths on disk we don't allow any exclusive actions (rename, drop so on)
+                /// until startup finished.
+                auto table_lock_holder = table->lockForShare(RWLockImpl::NO_QUERY, getContext()->getSettingsRef().lock_acquire_timeout);
+                table->startup();
+                logAboutProgress(log, ++tables_started, total_tables_to_startup, startup_watch);
+            }
+            else
+                throw Exception(ErrorCodes::UNKNOWN_TABLE, "Table {}.{} doesn't exist during startup",
+                    backQuote(name.database), backQuote(name.table));
+        });
 
-    try
-    {
-        for (const auto & table : TSA_SUPPRESS_WARNING_FOR_READ(tables))
-            thread_pool.scheduleOrThrowOnError([&]() { startup_one_table(table.second); });
-    }
-    catch (...)
+    return startup_table[name.table] = makeLoadTask(async_loader, {job});
+}
+
+LoadTaskPtr DatabaseOrdinary::startupDatabaseAsync(
+    AsyncLoader & async_loader,
+    LoadJobSet startup_after,
+    LoadingStrictnessLevel /*mode*/)
+{
+    std::scoped_lock lock(mutex);
+    // NOTE: this task is empty, but it is required for correct dependency handling (startup should be done after tables loading)
+    auto job = makeLoadJob(
+        std::move(startup_after),
+        AsyncLoaderPoolId::BackgroundStartup,
+        fmt::format("startup Ordinary database {}", database_name));
+    return makeLoadTask(async_loader, {job});
+}
+
+// TODO(serxa): implement
+// DatabaseTablesIteratorPtr DatabaseOrdinary::getTablesIterator(ContextPtr local_context, const DatabaseOnDisk::FilterByNameFunction & filter_by_table_name) const
+// }
+
+StoragePtr DatabaseOrdinary::tryGetTable(const String & name, ContextPtr local_context) const
```

Comment:
Do we also need to wait in `createTable`/`dropTable`/`detachTablePermanently`/`renameTable`/`alterTable`/etc?

Reply:
Yes, done